awx/main/models/notifications.py

```diff
@@ -68,6 +70,42 @@ class Meta:
 
     notification_configuration = JSONField(blank=False)
 
+    messages = JSONField(
+        null=True,
+        default={'started': { 'message': DEFAULT_MSG, 'body': ''},
+                 'success': { 'message': DEFAULT_MSG, 'body': ''},
+                 'error': { 'message': DEFAULT_MSG, 'body': ''}},
+        help_text=_('Optional custom messages for notification template.'))  # TODO: i18n?
+
+    def has_message(self, condition):
+        potential_template = self.messages.get(condition, {})
+        if potential_template == {}:
+            return False
+        if potential_template.get('message', {}) == {}:
+            return False
+        return True
+
+    def get_message(self, condition):
+        return self.messages.get(condition, {})
+
+    def build_notification_message(self, event_type, context):
+        env = sandbox.ImmutableSandboxedEnvironment()  # TODO: sub-class to define unsafe attrs
```

Comment:
Do these `TODOs` still need more code?

Reply:
I actually think that `ImmutableSandboxedEnvironment` might be fine out of the box since we're whitelisting acceptable fields. Dropping the todo.